lib/src/migrators/module/references.dart

```diff
@@ -244,18 +245,69 @@ class _ReferenceVisitor extends RecursiveAstVisitor {
     var stylesheet = result.item2;
     var canonicalUrl = stylesheet.span.sourceUrl;
     if (!_moduleScopes.containsKey(canonicalUrl)) {
+      var previousScope = _scope;
       _scope = Scope();
       _moduleScopes[canonicalUrl] = _scope;
       var oldImporter = _importer;
       _importer = result.item1;
       visitStylesheet(stylesheet);
       _checkUnresolvedReferences(_scope);
+      _scope = previousScope;
       _importer = oldImporter;
     }
     var namespace = namespaceForPath(node.url.path);
     _namespaces[namespace] = canonicalUrl;
   }
 
+  /// Visits the stylesheet this `@forward` rule points to using a new global
+  /// scope, then copies members from it into the current scope.
+  @override
+  void visitForwardRule(ForwardRule node) {
+    super.visitForwardRule(node);
+    var result = importCache.import(node.url, _importer, _currentUrl);
+    if (result == null) return;
+    var stylesheet = result.item2;
+    var canonicalUrl = stylesheet.span.sourceUrl;
+    if (!_moduleScopes.containsKey(canonicalUrl)) {
+      var previousScope = _scope;
+      _scope = Scope();
+      _moduleScopes[canonicalUrl] = _scope;
+      var oldImporter = _importer;
+      _importer = result.item1;
+      visitStylesheet(stylesheet);
+      _checkUnresolvedReferences(_scope);
+      _scope = previousScope;
+      _importer = oldImporter;
+    }
+    var moduleScope = _moduleScopes[canonicalUrl];
+    var prefix = node.prefix ?? '';
+    for (var name in moduleScope.variables.keys) {
+      if (moduleScope.variables[name] is! VariableDeclaration) {
+        throw StateError(
+            "Arguments should not be present in a module's global scope");
+      }
+      if ((node.shownVariables?.contains(name) ?? true) &&
+          !(node.hiddenVariables?.contains(name) ?? false)) {
```

Comment:
I'd make a helper function for this logic, something like `bool _visibleThroughForward(String name, Set<String> shown, Set<String> hidden)`.